#**Maestría en Inteligencia Artificial Aplicada**
##**Curso: Inteligencia Artificial y Aprendizaje Automático**
###Tecnológico de Monterrey
###Prof Luis Eduardo Falcón Morales

## **Adtividad de la Semana 10**
###**Reducción de Dimensionalidad - Descomposición en Valores Singulares (SVD) y Sistemas de Recomendación**


**Nombres y matrículas de los integrantes del equipo:**

*   
*   
*   
*   
*   


En cada sección deberás incluir todas las líneas de código necesarias para responder a cada uno de los ejercicios.

In [1]:
import numpy as np
import pandas as pd

from sklearn.decomposition import TruncatedSVD


#**Ejercicio-1.** 

1. Descarga los archivos y genera la matriz de utilidad en la cual los renglones son los usuarios (user_ID), las columnas los nombres de los restaurantes (name) y las entradas la evaluaciones generales dadas por los usuarios (rating). En particular indica el tamaño de dicha matriz y despliega sus primeros 10 renglones.

In [3]:
url1='/Users/alexishema/Downloads/RCdata/geoplaces2.csv'
url2='/Users/alexishema/Downloads/RCdata/rating_final.csv'

In [9]:
geo = pd.read_csv(url1,  sep=",",encoding='latin-1')  
ratings = pd.read_csv(url2,  sep=",",encoding='latin-1') 


In [16]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1161 entries, 0 to 1160
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   userID          1161 non-null   object
 1   placeID         1161 non-null   int64 
 2   rating          1161 non-null   int64 
 3   food_rating     1161 non-null   int64 
 4   service_rating  1161 non-null   int64 
dtypes: int64(4), object(1)
memory usage: 45.5+ KB


In [11]:
geo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130 entries, 0 to 129
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   placeID         130 non-null    int64  
 1   latitude        130 non-null    float64
 2   longitude       130 non-null    float64
 3   the_geom_meter  130 non-null    object 
 4   name            130 non-null    object 
 5   address         130 non-null    object 
 6   city            130 non-null    object 
 7   state           130 non-null    object 
 8   country         130 non-null    object 
 9   fax             130 non-null    object 
 10  zip             130 non-null    object 
 11  alcohol         130 non-null    object 
 12  smoking_area    130 non-null    object 
 13  dress_code      130 non-null    object 
 14  accessibility   130 non-null    object 
 15  price           130 non-null    object 
 16  url             130 non-null    object 
 17  Rambience       130 non-null    obj

In [19]:
matriz_ratings_grl = ratings[[ 'userID','rating','placeID']] 
matriz_geo_grl = geo[['name', 'placeID']] 



In [20]:
matriz_ratings_grl

userID  rating  placeID
0     U1077       2   135085
1     U1077       2   135038
2     U1077       2   132825
3     U1077       1   135060
4     U1068       1   135104
...     ...     ...      ...
1156  U1043       1   132630
1157  U1011       1   132715
1158  U1068       1   132733
1159  U1068       1   132594
1160  U1068       0   132660

[1161 rows x 3 columns]

In [21]:
matriz_geo_grl

name  placeID
0                   Kiku Cuernavaca   134999
1                   puesto de tacos   132825
2        El Rincón de San Francisco   135106
3    little pizza Emilio Portes Gil   132667
4                     carnitas_mata   132613
..                              ...      ...
125                         Chaires   132866
126                      Sushi Itto   135072
127                       Paniroles   135109
128     Restaurant Bar Coty y Pablo   135019
129                sirloin stockade   132877

[130 rows x 2 columns]

In [22]:
tabla_relacional = pd.merge(matriz_geo_grl,matriz_ratings_grl, on='placeID')   # dataFrame
tabla_relacional 

name  placeID userID  rating
0                 Kiku Cuernavaca   134999  U1093       2
1                 Kiku Cuernavaca   134999  U1066       1
2                 Kiku Cuernavaca   134999  U1040       1
3                 Kiku Cuernavaca   134999  U1110       2
4                 Kiku Cuernavaca   134999  U1121       2
...                           ...      ...    ...     ...
1156  Restaurant Bar Coty y Pablo   135019  U1121       2
1157             sirloin stockade   132877  U1047       0
1158             sirloin stockade   132877  U1111       1
1159             sirloin stockade   132877  U1015       1
1160             sirloin stockade   132877  U1025       1

[1161 rows x 4 columns]

In [23]:
UtMx = tabla_relacional.pivot_table(values='rating', index='userID', columns='name', fill_value=0)

UtMx.head(10)

name    Abondance Restaurante Bar  Arrachela Grill  Cabana Huasteca  \
userID                                                                
U1001                           0                0                0   
U1002                           0                0                0   
U1003                           0                0                2   
U1004                           0                0                0   
U1005                           0                0                0   
U1006                           0                0                0   
U1007                           0                0                0   
U1008                           1                0                0   
U1009                           0                0                1   
U1010                           0                0                0   

name    Cafe Chaires  Cafeteria cenidet  Cafeteria y Restaurant El Pacifico  \
userID                                                                        
U1001              0                  0                                   0   
U1002              0                  0                                   0   
U1003              0                  0                                   0   
U1004              0                  0                                   2   
U1005              0                  0                                   2   
U1006              1                  0                                   0   
U1007              1                  0                                   1   
U1008              0                  0                                   0   
U1009              0                  0                                   0   
U1010              0                  0                                   0   

name    Carls Jr  Carnitas Mata  Calle 16 de Septiembre  \
userID                                                    
U1001          0                                      0   
U1002          0                                      0   
U1003          0                                      0   
U1004          0                                      0   
U1005          0                                      0   
U1006          0                                      0   
U1007          0                                      0   
U1008          0                                      0   
U1009          0                                      0   
U1010          0                                      0   

name    Carreton de Flautas y Migadas  Cenaduria El Rincón de Tlaquepaque  \
userID                                                                       
U1001                               0                                    0   
U1002                               0                                    0   
U1003                               0                                    0   
U1004                               0                                    0   
U1005                               0                                    0   
U1006                               0                                    0   
U1007                               0                                    0   
U1008                               0                                    0   
U1009                               0                                    0   
U1010                               0                                    0   

name    ...  puesto de tacos  rockabilly   shi ro ie  sirloin stockade  \
userID  ...                                                              
U1001   ...                2            0          0                 0   
U1002   ...                2            0          0                 0   
U1003   ...                2            2          0                 0   
U1004   ...                0            0          0                 0   
U1005   ...                0            0          0                 0   
U1006   ...                1      

In [24]:
UtMx.shape 

(138, 129)

In [52]:
print('Total de elementos de la matriz de utilidad: %d' % (UtMx.size))
print('Total de elementos diferentes de cero: %d' % (np.count_nonzero(UtMx)))
print('Porcentaje de elementos diferentes de cero: %.1f%%' % (100 * np.count_nonzero(UtMx) / UtMx.size))

Total de elementos de la matriz de utilidad: 17802
Total de elementos diferentes de cero: 907
Porcentaje de elementos diferentes de cero: 5.1%


#**Ejercicio-2.**

2. Con el objetivo de obtener la información con respecto a los vectores latentes de los restaurantes, obtener dicha factorización SVD utilizando todas las componentes posibles, es decir, usando todos sus valores singulares. En particular indica la cantidad de valores singulares obtenidos.

In [215]:
X = UtMx.T     # Tomamos la traspuesta para que los renglones sean las películas y las columnas los usuarios.
X.shape

(129, 138)

In [220]:
SVD = TruncatedSVD(n_components=137, random_state=42)  
SVD.fit(X)


TruncatedSVD(n_components=137, random_state=42)

In [221]:
b=0
for a in SVD.singular_values_:
    b=b+1
print("La cantidad de valores singulares obtenidos es de:",b)

La cantidad de valores singulares obtenidos es de: 129


In [222]:
resultant_matrix = SVD.fit_transform(X)
resultant_matrix.shape

(129, 129)

#**Ejercicio-3.**

3. Usando la métrica de correlación de Pearon, obtener las 10 mejores recomendaciones de este modelo no supervisado con base a la información de alguien que evaluó bien de manera general al “Subway”. Despliega también el valor numérico de la correlación de Pearson de cada una de estas 10 recomendaciones. NOTA: No debe estar incluido “Subway” dentro de las recomendaciones.

In [223]:
corrMx = np.corrcoef(resultant_matrix, )
corrMx.shape

(129, 129)

In [224]:
me_gusta = "Subway"

names = UtMx.columns
names_list = list(names)
id_megusta = names_list.index(me_gusta)

id_megusta


88

In [225]:
corr_recomienda = corrMx[id_megusta]
print(corr_recomienda.shape)
print(corr_recomienda)

(129,)
[-7.69028369e-04 -7.87822168e-05 -5.27330975e-03  8.58511349e-04
 -3.03891148e-03  3.97495832e-05 -1.54820208e-03 -5.29347145e-03
  7.27520383e-04 -1.55098452e-03 -1.84483094e-03  3.01090491e-01
 -3.06938616e-05 -3.66639157e-03 -2.28024815e-03 -1.70971180e-03
  2.66961340e-03  4.58467365e-04 -2.11449533e-03 -6.20445134e-03
  3.34268563e-03  1.16598607e-01  1.87080102e-03  2.53682556e-03
 -5.48801141e-03 -2.91288266e-03  2.02825342e-03 -1.36607265e-03
 -2.01695899e-03  1.12058244e-01 -2.36235727e-03  6.52159239e-04
  3.53495770e-03 -3.62793330e-03  1.69065407e-03 -4.51837401e-03
 -2.09301342e-03  2.34634298e-01 -3.80413494e-03 -5.11845710e-04
 -2.27148704e-03  3.88178002e-01 -2.07427108e-03 -2.29428835e-03
  1.32565455e-03  3.01002479e-01  1.71433518e-01 -3.29715354e-03
 -4.74480596e-03 -8.02527884e-04 -2.65071452e-03  1.12018889e-01
 -3.91080724e-03 -2.06369812e-03  1.16008629e-01  2.27855804e-01
  4.28342495e-04 -8.69739761e-04  6.09520875e-01 -6.43364064e-03
  4.25467169e-01 -

In [226]:
corr_recomienda[id_megusta]

1.0

In [227]:
mejores10 = list(names[(corr_recomienda > 0.23)& (corr_recomienda < 1)])
mejores10

['Chilis Cuernavaca',
 'Log Yin',
 'Mariscos Tia Licha',
 'Mikasa',
 'Restaurant Teely',
 'Restaurant and Bar and Clothesline Carlos N Charlies',
 'Restaurant los Pinos',
 'Rincon del Bife',
 'Sanborns Casa Piedra',
 'Vips']

In [228]:
list(corr_recomienda[(corr_recomienda > 0.23)& (corr_recomienda < 1)])

[0.301090491044961,
 0.23463429773399852,
 0.3881780018525816,
 0.3010024785497628,
 0.6095208746935301,
 0.4254671689089249,
 0.28514921973296703,
 0.4298238308559109,
 0.41721850134793703,
 0.5062776799761117]

#**Ejercicio-4.**

4. Disminuye ahora la dimensión de la factorización con un número de componentes tal, que las recomendaciones con esta menor cantidad de información coincidan en al menos 9 de las 10 recomendaciones dadas con todos los valores singulares. Indica la cantidad de valores singulares encontrado y la cantidad de información que contienen.

In [230]:
def busqueda_SVD(n,X,id_megusta,mejores10): 
    
    SVD = TruncatedSVD(n_components=137, random_state=1)  
    SVD.fit(X)
    
    
    SVD_busqueda = TruncatedSVD(n_components=n, random_state=42) 
    SVD_busqueda.fit(X)
    resultant_matrix_busqueda= SVD_busqueda.fit_transform(X)
    corrMx_busqueda = np.corrcoef(resultant_matrix_busqueda, )
    corr_recomienda_busqueda = corrMx_busqueda[id_megusta]
    comparacion_mejores = list(names[(corr_recomienda_busqueda > 0.999955)& (corr_recomienda_busqueda < 1)])
    print("Longitud de la segunda lista de recomendaciones:",len(comparacion_mejores),"\n")
    print("Número de recomendaciones que son iguales entre la lista con valores singulares full y la segunda lista:",len(set(mejores10).intersection(comparacion_mejores)))
    print('\nCantidad de información simplificada con los primeros %d vectores singulares:' % n)
    print('%.1f%%' %  (100 * (1- (SVD.singular_values_[0:n]).sum() / (SVD.singular_values_).sum())))
    return 

busqueda_SVD(8,X,id_megusta,mejores10)

Longitud de la segunda lista de recomendaciones: 10 

Número de recomendaciones que son iguales entre la lista con valores singulares full y la segunda lista: 9

Cantidad de información simplificada con los primeros 8 vectores singulares:
79.3%


#**Ejercicio-5.**

5. Repite el ejercicio con el restaurante “"tacos de barbacoa enfrente del Tec". Incluye tus conclusiones.

In [231]:
me_gusta2 = "tacos de barbacoa enfrente del Tec"


id_megusta2 = names_list.index(me_gusta2)

id_megusta2

124

In [232]:
corr_recomienda2 = corrMx[id_megusta2]
print(corr_recomienda2.shape)
print(corr_recomienda2)

(129,)
[ 1.90365985e-03  1.95018219e-04  1.30535991e-02 -2.12516683e-03
  7.52254923e-03 -9.83964813e-05  3.83243356e-03  1.31035076e-02
  4.07794920e-01  3.83932123e-03  4.56671136e-03  1.56152873e-03
  7.59798653e-05  9.07581917e-03  5.64454708e-03  4.23223620e-03
 -6.60838540e-03 -1.13489430e-03  5.23424104e-03  1.53585555e-02
 -8.27451453e-03  4.82434184e-03 -4.63099792e-03 -6.27968115e-03
  1.48442849e-01  1.91120451e-01 -5.02075704e-03  3.38158871e-03
  4.99279870e-03  5.56810594e-03  5.84780076e-03 -1.61436094e-03
 -8.75046657e-03  8.98061922e-03 -4.18506052e-03  1.11848243e-02
  2.58479385e-01  1.27230718e-02  9.41679037e-03  1.26702755e-03
  5.62285977e-03  5.74403685e-03  5.13466958e-03  5.67930233e-03
 -3.28153738e-03  1.83689220e-03  1.11393805e-02  8.16180397e-03
  1.17453360e-02  1.98658485e-03  6.56160293e-03  5.69993965e-03
  9.68084797e-03  5.10849717e-03  6.85473722e-03 -3.44059875e-03
 -1.06032292e-03  2.15296175e-03  1.33109496e-04  1.59258928e-02
  6.48831289e-03  

In [243]:
mejores10_2 = list(names[(corr_recomienda2 > 0.1709)& (corr_recomienda2 < 1)])
mejores10_2

['Carreton de Flautas y Migadas',
 'Hamburguesas La perica',
 'Little Cesarz',
 'Taqueria EL amigo ',
 'carnitas mata calle Emilio Portes Gil',
 'little pizza Emilio Portes Gil',
 'palomo tec',
 'puesto de gorditas',
 'tacos abi',
 'vips']

In [245]:
list(corr_recomienda2[(corr_recomienda2 > 0.1709)& (corr_recomienda2 < 1)])

[0.40779492018129593,
 0.19112045068235076,
 0.258479385426264,
 0.4371842084263567,
 0.17093947240845903,
 0.6538492684416535,
 0.19517408801078595,
 0.41276452318483897,
 0.6653053654632715,
 0.8171889849105266]

In [439]:
def busqueda_SVD_2(n,X,id_megusta,mejores10): 
    
    SVD = TruncatedSVD(n_components=137, random_state=1)  
    SVD.fit(X)
    
    
    SVD_busqueda = TruncatedSVD(n_components=n, random_state=42) 
    SVD_busqueda.fit(X)
    resultant_matrix_busqueda= SVD_busqueda.fit_transform(X)
    corrMx_busqueda = np.corrcoef(resultant_matrix_busqueda, )
    corr_recomienda_busqueda = corrMx_busqueda[id_megusta]
    comparacion_mejores = list(names[(corr_recomienda_busqueda > 0.175)& (corr_recomienda_busqueda < 1)])
    print("Longitud de la segunda lista de recomendaciones:",len(comparacion_mejores),"\n")
    print("Número de recomendaciones que son iguales entre la lista con valores singulares full y la segunda lista:",len(set(mejores10).intersection(comparacion_mejores)))
    print('\nCantidad de información simplificada con los primeros %d vectores singulares:' % n)
    print('%.1f%%' %  (100 * (1- (SVD.singular_values_[0:n]).sum() / (SVD.singular_values_).sum())))
    return 

busqueda_SVD_2(79,X,id_megusta2,mejores10_2)

Longitud de la segunda lista de recomendaciones: 10 

Número de recomendaciones que son iguales entre la lista con valores singulares full y la segunda lista: 9

Cantidad de información simplificada con los primeros 79 vectores singulares:
9.9%


**Conclusiones del Ejercicio 5** - La búsqueda para encontrar valores singulares reducidos vs el ejercicio 4 fue muy diferente. Se hicieron ejercicios manuales y para los componentes 8, prácticamente ya habíamos conseguido el 90% de similitud. Intentamos aplicar la misma metodología para el ejercicio 5 y apenas para los componentes 79 es que pudimos igualar la misma eficiencia. En este caso la información simplificada fue muy poca (9.9% comparado con 79.3% del otro ejercicio) lo cual prácticamente nos dice que necesitamos casi la misma información para poder tener ese grado de precisión.  

#**Ejercicio-6.**

6. Repite los ejercicios 1 a 4 generando ahora la matriz de utilidad en la cual los renglones son los usuarios (user_ID), las columnas los nombres de los restaurantes (name) y las entradas la evaluación dada a la comida (food_rating). Incluye tus conclusiones.

In [ ]:
None

#**Ejercicio-7.**

In [ ]:
None

#**Ejercicio-8.**

In [ ]:
None

###**Fin de la Actividad de la semana 10 - Sistemas de Recomendación y Reducción de Dimensionalidad con SVD.**